### PR TITLE
zlib: 1.2.11 -> 1.2.12 (security, CVE-2018-25032)

### PR DESCRIPTION
**Pull request by wamserma**
This version bump is the official fix for CVE-2018-25032.

Release Notes:
https://zlib.net/

More info on CVE:
https://nvd.nist.gov/vuln/detail/CVE-2018-25032
https://www.openwall.com/lists/oss-security/2022/03/24/1

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/zlib/default.nix [hunk 1]
```diff
@@ -23,23 +23,17 @@ assert splitStaticOutput -> static;
...
-  version = "1.2.11";
+  version = "1.2.12";
 
   src = fetchurl {
     urls =
       [ "https://www.zlib.net/fossils/zlib-${version}.tar.gz"  # stable archive path
         "mirror://sourceforge/libpng/zlib/${version}/zlib-${version}.tar.gz"
       ];
-    sha256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1";
+    sha256 = "c3e5e9fdd522222b542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1";
```
Comment (numinit) [inline pkgs/development/libraries/zlib/default.nix:33]: What is the SHA256  `c3e5e9fdd522222b542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1`? It looks like the old SHA256 `c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1` with a few bits corrupted.

The actual SHA256 of zlib should be  `91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9`.

**Review by numinit — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/zlib/default.nix (hunk 1 above)
Comment (wamserma) [reply pkgs/development/libraries/zlib/default.nix]: Ah, pushed to early. Fixed.

**Comment by numinit** [comment]
@GrahamcOfBorg eval
@GrahamcOfBorg build zlib

File: pkgs/development/libraries/zlib/default.nix [hunk 2]
```diff
@@ -23,23 +23,17 @@ assert splitStaticOutput -> static;
...
-  version = "1.2.11";
+  version = "1.2.12";
 
   src = fetchurl {
     urls =
       [ "https://www.zlib.net/fossils/zlib-${version}.tar.gz"  # stable archive path
         "mirror://sourceforge/libpng/zlib/${version}/zlib-${version}.tar.gz"
       ];
-    sha256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1";
+    sha256 = "1n9na4fq4wagw1nzsfjr6wyly960jfa94460ncbf6p1fac44i14i";
```
Comment (numinit) [inline pkgs/development/libraries/zlib/default.nix:33]: Maybe it's a better idea to use the hex representation? This is the right hash, but the hex representation of a hash for a package that rarely changes is easier to verify at a glance against the zlib website (which is where I got that hash)

Comment (wamserma) [reply pkgs/development/libraries/zlib/default.nix]: Good point. Switched to base16.

File: pkgs/development/libraries/zlib/default.nix [hunk 3]
```diff
@@ -23,23 +23,17 @@ assert splitStaticOutput -> static;
...
-  patches = [
-    # https://nvd.nist.gov/vuln/detail/CVE-2018-25032
-    # https://github.com/madler/zlib/commit/5c44459c3b28a9bd3283aaceab7c615f8020c531
-    ./CVE-2018-25032-1.patch
-    # https://github.com/madler/zlib/commit/4346a16853e19b45787ce933666026903fb8f3f8
-    ./CVE-2018-25032-2.patch
-  ] ++ lib.optionals stdenv.hostPlatform.isCygwin [
+  patches = lib.optionals stdenv.hostPlatform.isCygwin [
     ./disable-cygwin-widechar.patch
   ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/zlib/default.nix:38]: I am force pushing the removal of this patch in a few minutes because it got upstreamed, too.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Please ping me if no one else merges this within a day or two.

**Comment by vcunat** [comment]
It rebuilds all stdenvs, so binaries won't be there so soon either way.  We might want to fix stable sooner than unstable?

As for unstable; I wasn't following the current staging-next iteration now (as I'm mostly unavailable this week), so perhaps it might be ready enough to go with this at once (say, merge this to staging-next and then soon to master when a few binaries are there).

**Comment by wamserma** [comment]
Shall I rebase the commit onto staging-next or do you prefer a cherry-pick?

**Comment by mweinelt** [comment]
Going via staging-next is only worth it if staging-next is actually in a good state, we have to look into that first.

**Comment by mweinelt** [comment]
> Shall I rebase the commit onto staging-next or do you prefer a cherry-pick? I'd suggest the cherry-pick ashttps://github.com/[/pull/165642](https://github.com/NixOS/nixpkgs/pull/165642) has already been merged to staging.

Picked that commit into staging-next, please rebase onto staging-next.

**Comment by vcunat** [comment]
- [ ] `qt5.qtbase` can no longer find zlib on x86_64-darwin.  Example 21.11 log: https://hydra.nixos.org/build/171602278

**Comment by wamserma** [comment]
> * [ ]  `qt5.qtbase` can no longer find zlib on x86_64-darwin.  Example 21.11 log: https://hydra.nixos.org/build/171602278

The Qt-bundled zlib was updated in
https://code.qt.io/cgit/qt/qtbase.git/commit/?id=a0a2bf2d95d4fcd468b6ce3c2e728d95425dd760 
which says `Pick-to: 6.2 6.3 6.3.0 5.15` but it seems there is no pick to any Qt5 version (least 5.12.10, which is in our stable).

**Comment by vcunat** [comment]
I hope we don't use the bundled one, but the commit (or some around it) might contain other necessary changes for newer zlib in Qt.  Darwin defaults to Qt 5.12 even on _newest nixpkgs_... so perhaps that would better get addressed some time, too.

**Comment by mweinelt** [comment]
Well the root cause might be similar to PyPy

- [ ] `pypy3` fails to find zlib. Example on staging-next: https://hydra.nixos.org/build/172082854

**Comment by Mindavi** [comment]
This broke support for cross-compiling zlib, causing everything depending on it to break:

```
mindavi@nixos:~/nixpkgs$ nix build .#pkgsCross.aarch64-multiplatform.zlib
mindavi@nixos:~/nixpkgs$ tree result
result
└── share
    └── man
        └── man3
            └── zlib.3.gz

3 directories, 1 file
mindavi@nixos:~/nixpkgs$ nix build .#pkgsCross.aarch64-multiplatform.zlib.dev
mindavi@nixos:~/nixpkgs$ tree result-dev
result-dev
├── include
│   ├── zconf.h
│   └── zlib.h
├── lib
│   └── pkgconfig
│       └── zlib.pc
└── nix-support
    └── propagated-build-inputs

4 directories, 4 files
```

Commit on which I tested this (merge staging-next to master): b4729bad3d1c1c78dbd47634a1efba388c89dbbb

Maybe this section in the log points to something?

```
zlib-aarch64-unknown-linux-gnu> Checking for shared library support...
zlib-aarch64-unknown-linux-gnu> No shared library support; try without defining CC and CFLAGS
zlib-aarch64-unknown-linux-gnu> Building static library libz.a version 1.2.12 with aarch64-unknown-linux-gnu-gcc.
```

**Pull request merged**